### ISSUE TITLE
Create a Resource type for audio spatialization models

### ISSUE THREAD

**Issue by ellenhp** [godotengine/godot-proposals#4377]
### Describe the project you are working on

Godot engine audio system

### Describe the problem or limitation you are having in your project

Users are requesting more advanced audio capabilities for XR, specifically HRTF's to better spatialize audio around the headset. This comes with a number of problems:

* We must retain surround sound support (at least 3.1 and 5.1) which means we must retain SPCAP. It's old but it's the most advanced irregular-format panning system I'm aware of. Many XR-only audio systems mix to a higher-order ambisonic format but decoding ambisonics to irregular speaker configurations is incredibly difficult.
* We can't allow callbacks into script from the audio thread (this would allow users to implement whatever they wanted for themselves, but would likely cause buffer underruns)
* We can't drastically overhaul the architecture of the audio system because we don't have enough audio contributors to go implement those changes and support them.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I thought about how to solve this in the most "Godot" way possible, and pretty quickly came to the conclusion that the HRTFs themselves do not belong in core, meaning Godot shouldn't ship with any specific data from this [dataset](https://www.ece.ucdavis.edu/cipic/spatial-sound/hrtf-data/). Each HRTF is about a megabyte or 700kb compressed. But I do like the idea of allowing a much more configurable `AudioStreamPlayer3D` experience for users, so I kept thinking.

I propose a new Resource type, called `SpatialAudioModel` that will be responsible for maintaining a list of HRTFs for different azimuths and elevations around the listener. These HRTFs could be measured, like the ones from the dataset described above, or they could be hand-crafted for the sole purpose of defining a specific interaural delay.

`SpatialAudioModel` could (?) also be responsible for maintaining an effect chain and defining which specific parameters to use for that effect chain at various distances, which will allow people to boost the bass on nearby sounds if they wish, or make whatever other adjustments are appropriate for their game and tastes.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

I think this should work similarly to bus layouts, in that there should be a project setting that anyone with an XR game can set to point to a specific `SpatialAudioModel` .tres file they've downloaded from the Asset Library and put somewhere in their project. On initialization, the `AudioServer` will load the spatial audio model if that project setting is set.

Once the spatial audio model is set in the `AudioServer`, any AudioStreamPlayer3D that is asked to begin playback will use that model unless it is set up specifically to fall back on SPCAP. It should also fall back on SPCAP if it is mixing to a channel layout other than stereo. When it falls back to SPCAP it would still be nice if it used the effect chains defined in the `SpatialAudioModel`, but I'm not quite sure how that will work architecturally yet.

I'm thinking that it makes the most sense to create a new `AudioStreamPlayback` type called `SpatialAudioStreamPlayback` that will do the job of applying the effect chain and HRTF from the spatial model. There are complicated reasons for this that I can go into in comments on this proposal if people are interested. They mostly relate to the fact that it's more performant to apply a dense convolution in the frequency domain.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No.

### Is there a reason why this should be core and not an add-on in the asset library?

It can't be worked around. Plus, I really like the idea of being able to let users download an entire new spatialization system for their game from the Asset Library.

**Comment by ellenhp** [comment]
Do we use different labels for proposals? Sorry if I mislabeled!

**Comment by Calinou** [comment]
> Do we use different labels for proposals? Sorry if I mislabeled!

Yes. In fact, I don't know how the `enhancement` label was added here, since every proposal is an enhancement request by definition.

**Comment by ellenhp** [comment]
I added it without thinking but that is a very good point. 😆

Anyway, this is the only HRTF proposal I've seen that seems feasible. I could (and am willing to) build this in a few weekends. My big questions are:

1. Does the Godot XR community want this, or will we be happy with regular stereo panning after https://github.com/godotengine/godot/pull/58841 is merged?
2. Does the core contributors think this approach makes sense? Audio mixing is all deep inside the core engine code and I want to allow users to change bits and pieces of it to fit their needs, but I don't want to massively re-architect the engine to do that, or compromise on other use-cases. This is the most lightweight thing I can think of, and non-XR users would probably never even notice it's there.

It's also probably one of the only ways to get good spatial audio for XR that doesn't break compat or change behavior for existing games.

**Comment by BastiaanOlij** [comment]
Can't really comment on the technical implementation, audio is a bit of a blind spot in my skillset at the moment (I should really correct that some day), but I'm looking forward to seeing this in action and definitely believe Godot (especially XR in Godot) will benefit from a proper spatial audio system.

**Comment by ellenhp** [comment]
Summarizing discussion from Proposal Review Meeting:

1. We think this feature is worth implementing in core.
2. We think that including the actual HRTF impulse responses in a resource (as opposed to an export template) is the correct approach.
3. We want to talk more about implementation in an audio contributor's meeting (open to all).

----- end summary ------

I also identified a problem before the meeting with the idea of wrapping `AudioStreamPlayback`s in a spatial playback. This would apply the HRTF *before* sending the audio to the bus system, which means audio effects applied by the bus system could destroy the spatial cues provided by the HRTF.

We might have to rethink the way the bus system works, IMO.

**Comment by ellenhp** [comment]
See also: https://github.com/godotengine/godot/issues/23485
